### ISSUE TITLE
Feature request: cap the size of the message queue

### ISSUE THREAD

**Issue by PaulFake** [chkr1011/MQTTnet#461]
Is it possible to add a configurable cap on the size of the message queue?  Currently, if I interrupt the connection between client and broker while continuously publishing a very large number of messages, the message queue will continue to grow indefinitely.  I would like to set a cap on it so I don't crash the program, possibly even with some kind of notification that the limit has been reached and further published messages will be discarded.

**Comment by PaulFake** [comment]
Yes

**Comment by chkr1011** [comment]
I released 2.8.5-rc1. Please let me know if version works for you.
Best regards
Christian

**Comment by PaulFake** [comment]
Awesome!  I'll check it out soon and get back to you - dealing with another thing right now before I can do that.

**Comment by PaulFake** [comment]
I've tested out the DropNewMessage strategy by setting the cap, then adding logging to report messages.Count inside my IManagedMqttClientStorage.SaveQueuedMessagesAsync implementation.  What I'm seeing is that I always get exactly 2 more messages than the cap.  For instance, if I set the cap to 5, I get a maximum of 7 before the OnApplicationMessageSkipped event is raised and SaveQueuedMessagesAsync is no longer called.  It even behaves the same when I set the cap to zero (it allows 2 messages in the queue).  If there's no explanation for why it should do this, I'll pull down the source and see what I can find debugging it.

**Comment by chkr1011** [comment]
OK that sounds strange. But avoid saving is OK because the internal list has not changed. Only new messages are truncated. This will not affect the queue at all. So I avoided useless calls for saving.

**Comment by chkr1011** [comment]
Hi I released 2.8.5-rc2. I found an issue which causes usually having one more pending message as configured.

**Comment by PaulFake** [comment]
I upgraded to rc2 and it is now capping at only one more than the cap (6, instead of 7, when configured with 5) - so that's definitely better.  I'm not particularly worried about this small difference, as our users will probably be capping in the thousands and they'll never notice the discrepancy.  During one of my runs, a message got stuck in the queue after coming back from a failed connection.  It never drained from the queue until I restarted the app.  I couldn't reproduce this after seeing it, but I'll keep an eye out and let you know if I can reproduce it reliably.

**Comment by PaulFake** [comment]
Yes, I am referring to the storage.  I'm logging the count of the parameter of SaveQueuedMessagesAsync. Sorry for not clarifying that.

**Comment by PaulFake** [comment]
For an update, we ended up using the DropOldestQueuedMessage strategy, as it made more sense for our purposes.  Our QA was testing it and saw that with a cap of 100, the storage would reach 101, which would then trigger the ApplicationMessageSkipped event.  Then after the storage drained back down to 0, it filled up to 800 (far surpassing the cap of 100), triggered the ApplicationMessageSkipped event, and then proceeded to continue filling to 7430, when it stopped draining until the app was restarted.

Should the storage queue have a one-to-one relationship with the message queue?  I always assumed they were the same thing, but now I'm not really clear on that.

**Comment by PaulFake** [comment]
Interestingly, the behavior I described above only happens for us when the cap is low - QA was able to reliably reproduce it with a cap of 100, but not with a cap of 5000.

**Comment by chkr1011** [comment]
Maybe this is related to async? The storage is not saving a full list of messages. Instead it has the _Add_ and _Remove_ methods. After these are called the list ist saved via _Save_. Is there maybe an issue in the storage implementation?

**Comment by PaulFake** [comment]
Hm...the implementation is pretty simple:

```
    class FileMqttNetClientStorage : MqttNetClientStorageBase
    {
        private readonly string Filename;
        private IList<ManagedMqttApplicationMessage> _messages = null;

        public FileMqttNetClientStorage(string path, ILogger logger)
            : base(logger)
        {
            Filename = path + "RetainedMessages.json";
        }

        public override Task SaveQueuedMessagesAsync(IList<ManagedMqttApplicationMessage> messages)
        {
            Logger.Debug($"{messages.Count} message(s) in the queue.");
            _messages = messages;
            return Task.FromResult(0);
        }

        public override Task<IList<ManagedMqttApplicationMessage>> LoadQueuedMessagesAsync()
        {
            if (File.Exists(Filename))
            {
                var json = File.ReadAllText(Filename);
                File.Delete(Filename);
                _messages = JsonConvert.DeserializeObject<List<ManagedMqttApplicationMessage>>(json);
            }
            else
            {
                _messages = new List<ManagedMqttApplicationMessage>();
            }

            return Task.FromResult(_messages);
        }

        public override void Dispose()
        {
            var filteredMessages = _messages.Where((x) => x.ApplicationMessage.QualityOfServiceLevel != MQTTnet.Protocol.MqttQualityOfServiceLevel.AtMostOnce).ToList();
            File.WriteAllText(Filename, JsonConvert.SerializeObject(filteredMessages));
        }
    }
```

Basically it just holds onto a reference to the list so it can dump it when the app exits.

**Comment by PaulFake** [comment]
I want to keep you updated on issues we're seeing with the cap.  Our QA team was able to reliably reproduce a bug in which a message gets stuck in the storage queue (never drains).  I am able to reproduce this bug on my own system as well.

1) Set the cap to 200,000
2) Enable DropOldestQueuedMessage
3) Run your app
4) Start Clumsy or some other tool to interrupt the connection to the broker
5) Before the Disconnected event is raised, publish two messages (which both end up in the storage queue)
6) Wait for the Disconnected event to raise, then stop Clumsy
7) When publishing resumes, the queue will drain but leave 1 item in it permanently

A couple of notes: strangely enough, the size of the cap matters.  Even though we don't actually reach the cap, setting it to a more conservative 5000 does not have the same result as 200,000.  Also, this could be some kind of race condition on our system, because our systems automatically publish a message when the connection is re-established, so there's an extra message being published between steps 6 and 7.  I've pulled down the code and I'm going to start debugging, so I might have this solved soon.  If I do, I'll try submitting the fix to the project.

**Comment by PaulFake** [comment]
[https://github.com/chkr1011/MQTTnet/pull/476](url)

**Comment by PaulFake** [comment]
The other issue should be fixed in #479.
